### PR TITLE
Extend GroundPrimitive shadow volume based on screen space error

### LINKED ISSUE #4161: Polygon triangles disappear when zoomed out far
missing-triangles

http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Polygon.html&label=Geometries
Hit home then continue to zoom out, you'll see triangles on the red and blue polygons disappear

Seems to have been introduced in 1.23

### LINKED ISSUE #4326: KML linestring clamp to ground fails
Reported on the forum: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/9wvnNNQcLTA
This KML renders fine when `clampToGround: false` but doesn't render anything when `clampToGround: true`.  Zooming to it then causes a crash in `Scene updateFrustums` because NaNs populate the frustum values.
[track.kml.txt](https://github.com/AnalyticalGraphicsInc/cesium/files/477029/track.kml.txt)

``` javascript
var viewer = new Cesium.Viewer('cesiumContainer');
var options = {
    camera : viewer.scene.camera,
    canvas : viewer.scene.canvas,
    clampToGround: true
};


viewer.zoomTo(viewer.dataSources.add(Cesium.KmlDataSource.load('../../SampleData/kml/track.kml', options)));
```

### PR DESCRIPTION
**Pull request by hpinkos**
Fixes #4326
Fixes #4161

This extends the bottom of the shadow volume below the terrain based on the screen space error.  The shadow volume is updated in the ShadowVolume vertex shader.

- Ground primitives geometry types compute a `extrudeDirection` attribute.  The `extudeDirection` is a normalized vector pointing in the direction in which the position should be moved in the vertex shader.  
  - Positions at the top of the shadow volume will have `extrudeDirection === Cartesian3.ZERO` because only the bottom positions need to be moved
- Added `czm_geometricToleranceOverDistance ` to `AutomaticUniforms`
- Added `pixelSizePerMeter` and `maximumScreenSpaceError` to `UniformState` 
  - `scene.globe.maximumScreenSpaceError` get piped through to `UniformState` via the `frameState`.  Not sure if that's the best way to do this.
- Added `u_globeMinimumAltitude` to `GroundPrimitive.uniforms`.  
  - I set this to `33000.0`, which is a magic number I found with testing.  I'm not sure what this value should actually be. 
  - I made this a positive number so I don't have to `abs` it later when i find the `min` between this value and the value computed using the screen space error.

TODO
- [x] Fix debugShowShadowVolume so it displays the shadow volume after the positions have been updated in the shader
- [x] octencode `extrudeDirection` attribute
- [x] Update CHANGES.md

### CODE REVIEW

**Comment by pjcozzi** [comment]
> Ground primitives geometry types compute a extrudeDirection attribute.

Let's octencode this.  See [AttributeCompression.js](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/AttributeCompression.js).

**Comment by pjcozzi** [comment]
> I set this to 33000.0, which is a magic number I found with testing. I'm not sure what this value should actually be.

This should come from the terrain provider, but we can hard code it to the right min for Earth for now.  @bagnell do you have it handy from STK?

**Comment by pjcozzi** [comment]
> Positions at the top of the shadow volume will have extrudeDirection === Cartesian3.ZERO because only the bottom positions need to be moved

@lilleyse @bagnell do you guys have any ideas on how to avoid this wasted memory without introducing another draw call per volume?

**Comment by pjcozzi** [comment]
Update CHANGES.md.

**Comment by pjcozzi** [comment]
Perhaps there is a winding order issue:

image

Click "Create large polygons"

http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/ground-primitive-sse/Apps/Sandcastle/index.html?src=development/Ground%20Primitive.html&label=Showcases

**Comment by pjcozzi** [comment]
>> Positions at the top of the shadow volume will have extrudeDirection === Cartesian3.ZERO because only the bottom positions need to be moved

> @lilleyse @bagnell do you guys have any ideas on how to avoid this wasted memory without introducing another draw call per volume?

Related idea: at one point, I just tried to normalize the WGS84 position, to avoid storing an extra attribute, and it was very noticeably wrong at places like Mount Saint Helens.  However, perhaps there is another way to compute a better approximation, but still not perfect.

**Comment by pjcozzi** [comment]
Also with that same example, the corridor and circle disappear when zoomed out.

bv

**Comment by pjcozzi** [comment]
@hpinkos did you try using `GroundPrimitive.debugShowShadowVolume` to verify that the volumes are as low as they need to be (so it renders correctly), but not much lower (to avoid fillrate)?

File: Source/Renderer/AutomaticUniforms.js [hunk 1]
```diff
@@ -1476,6 +1476,20 @@ define([
...
             }
+        }),
+
+        /**
+         * An automatic GLSL uniform scalar representing the geometric tolerance per distance
+         *
+         * @alias czm_geometricToleranceOverDistance
+         * @glslUniform
+         */
+        czm_geometricToleranceOverDistance : new AutomaticUniform({
```
Comment (pjcozzi) [inline Source/Renderer/AutomaticUniforms.js]: `Distance` -> `Meter`

File: Source/Renderer/UniformState.js [hunk 2]
```diff
@@ -935,6 +961,14 @@ define([
 
         this._frameState = frameState;
         this._temeToPseudoFixed = Transforms.computeTemeToPseudoFixedMatrix(frameState.time, this._temeToPseudoFixed);
+
+        var fov = camera.frustum.fov;
+        var viewport = this._viewport;
+        if (viewport.width > viewport.height * camera.frustum.aspectRatio) {
```
Comment (pjcozzi) [inline Source/Renderer/UniformState.js]: How does this check work?

File: Source/Shaders/ShadowVolumeVS.glsl [hunk 3]
```diff
@@ -17,7 +20,11 @@ vec4 depthClampFarPlane(vec4 vertexInClipCoordinates)
 void main()
 {
     v_color = color;
-    
+
     vec4 position = czm_computePosition();
+    float delta = min(u_globeMinimumAltitude, czm_geometricToleranceOverDistance * length(position.xyz));
+
+    //extrudeDirection is zero for the top layer
+    position = position + vec4(extrudeDirection.xyz * delta, 0);
```
Comment (pjcozzi) [inline Source/Shaders/ShadowVolumeVS.glsl]: Use `0.0`, mixing integers and floats didn't use to compile on some WebGL implementations.

Comment (pjcozzi) [inline Source/Shaders/ShadowVolumeVS.glsl]: `extrudeDirection` is a `vec3` so we can remove the `.xyz `.

**Comment by pjcozzi** [comment]
What do you think about cleaning up the geometry types by adding adding a temporary `BOTTOM` (or whatever) attribute then running the `Geometry` through a generic function that removes that attribute and adds the extruded normal?  It would come at a potential small performance and GC cost but could be worth it.

@bagnell what do you think about this for vector tiles?  Seems like we don't need to store the extruded direction in the payload.  We could even separate out the bottom cap vertices so the `BOTTOM` attribute is implicit over the wire.

I feel like we should do some fill rate tests before getting into all this.  This should make a huge difference for horizon and 45 degree views depending on how well early-z does (even with early-z the rasterization load is very high).

File: Source/Renderer/UniformState.js (hunk 2 above)
Comment (hpinkos) [inline Source/Renderer/UniformState.js]: This was just a straight port.  Did I misinterpreted something? Here's the original source:
``` c
    if ((double)mainViewport[3] > ((double)mainViewport[2] * p3dScene->pixelAspect))
    {
        pixelSizePerDistance = tan(0.5 * p3dScene->view.m_pCamera->FieldOfView()) * 2.0 / (double)mainViewport[3];
    }
    else
    {
        pixelSizePerDistance = tan(0.5 * p3dScene->view.m_pCamera->FieldOfView()) * 2.0 / (double)mainViewport[2];
    }
```

**Comment by hpinkos** [comment]
Here is a visualization of the shadow volume.  The blue is the old volume and the red is the new one.  It just extends a little bit below what we were doing previously

large polygon:
image

small rectangle:
image

**Comment by hpinkos** [comment]
@pjcozzi I believe I addressed your comments and fixed the problems you found with some of the geometries.

I haven't done the octencodeing yet because I'm waiting to see if we want to change the `extrudeDirection` attribute

**Comment by pjcozzi** [comment]
> Fix debugShowShadowVolume so it displays the shadow volume after the positions have been updated in the shader

Added in 2d50df1ef5a9c878e5fd75020fc58c4bcc8c9191.

This restriction was removed:

>Must be <code>true</code> on creation for the volumes to be created before the geometry is released or releaseGeometryInstances must be <code>false</code>

But the polygon is no longer drawn when the volume is drawn.

**Comment by pjcozzi** [comment]
> I haven't done the octencodeing yet because I'm waiting to see if we want to change the extrudeDirection attribute

Let's just move forward so we don't stall this.  We can optimize later if we come up with a good idea.

File: Source/Renderer/UniformState.js (hunk 2 above)
Comment (pjcozzi) [inline Source/Renderer/UniformState.js]: Are you sure `mainViewport[3]` is `width`?  In Cesium, the array would be `[x, y, width, height]`.

Are you sure `pixelAspect` ratio is `width / height` like `aspectRatio`?

Perhaps ask Greg to double check these.

Right now, I think the `if` statements reads like "if (width > (height * (width / height))", which is "if (width > width)"

File: Source/Renderer/AutomaticUniforms.js [hunk 4]
```diff
@@ -1476,6 +1476,20 @@ define([
...
+         * An automatic GLSL uniform scalar representing the geometric tolerance per meter
+         *
+         * @alias czm_geometricToleranceOverMeter
+         * @glslUniform
+         */
+        czm_geometricToleranceOverMeter : new AutomaticUniform({
+            size: 1,
+            datatype: WebGLConstants.FLOAT,
+            getValue: function(uniformState) {
+                return uniformState.pixelSizePerMeter * uniformState.maximumScreenSpaceError;
```
Comment (pjcozzi) [inline Source/Renderer/AutomaticUniforms.js]: This is constant per frame, right?  Precompute and store in `UniformState`.

File: Source/Renderer/UniformState.js [hunk 5]
```diff
@@ -768,6 +770,30 @@ define([
         },
 
         /**
+         * The maximum screen-space error used to drive level-of-detail refinement.  Higher
+         * values will provide better performance but lower visual quality.
+         *
+         * @type {Number}
+         * @default 2
+         */
+        maximumScreenSpaceError: {
```
Comment (pjcozzi) [inline Source/Renderer/UniformState.js]: These two getters may not be needed after the about change.

**Comment by pjcozzi** [comment]
Just those comments.

**Comment by hpinkos** [comment]
@pjcozzi this is ready for another look.  I fix the issues you pointed out and I added octencode for the new attribute.  And I updated `CHANGES.md`

**Comment by bagnell** [comment]
Oct-encoding is fine for lighting because the error isn't noticeable but it may be noticeable for geometry extruded with these normals. Have you seen this at all?

**Comment by hpinkos** [comment]
I didn't see any noticeable errors, but I'll look more closely to make sure

**Comment by pjcozzi** [comment]
> Oct-encoding is fine for lighting because the error isn't noticeable but it may be noticeable for geometry extruded with these normals

How many bits are we using?  It should be OK if we use enough....more than lighting.  For example, normals for rotation of instances in i3dm use 32: https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/TileFormats/Instanced3DModel#oct-encoded-normal-vectors

The paper has exact error bounds.

**Comment by pjcozzi** [comment]
Bump:

>> Oct-encoding is fine for lighting because the error isn't noticeable but it may be noticeable for geometry extruded with these normals
> How many bits are we using? It should be OK if we use enough....more than lighting. For example, normals for rotation of instances in i3dm use 32: https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/TileFormats/Instanced3DModel#oct-encoded-normal-vectors

> The paper has exact error bounds.

**Comment by hpinkos** [comment]
Thanks @bagnell, @pjcozzi, I'll update it to use 32

**Comment by hpinkos** [comment]
@pjcozzi @bagnell ready

File: Source/Shaders/Builtin/Functions/octDecode.glsl [hunk 6]
```diff
@@ -1,3 +1,29 @@
...
+  * Decodes a unit-length vector in 'oct' encoding to a normalized 3-component Cartesian vector.
+  * The 'oct' encoding is described in "A Survey of Efficient Representations of Independent Unit Vectors",
+  * Cigolle et al 2014: http://jcgt.org/published/0003/02/01/
+  *
+  * @name czm_octDecode
+  * @param {vec2} encoded The oct-encoded, unit-length vector
+  * @param {float} range The maximum value of the SNORM range. The encoded vector is stored in log2(rangeMax+1) bits.
+  * @returns {vec3} The decoded and normalized vector
+  */
+  vec3 czm_octDecode(vec2 encoded, float range)
```
Comment (pjcozzi) [inline Source/Shaders/Builtin/Functions/octDecode.glsl]: I'll be pleasantly surprised if JSDoc handles function overloads, but it's not an issue now.

**Comment by pjcozzi** [comment]
Do you get these artifacts?  If so, do you also get them without normal compression?

ok2

If these are due to normal compression, is there anything we can do about it?  Are we sure the code is right?  Is there a more precise (48 bits?) version from the paper we can try?

**Comment by hpinkos** [comment]
@pjcozzi no I don't see that on my machine.  That's happening after the 32P update?

**Comment by pjcozzi** [comment]
I don't get this on Windows.  Will test Mac/Intel now, hopefully the artifact is in master.

**Comment by pjcozzi** [comment]
Good news, it is in master.  See #4820

**Comment by pjcozzi** [comment]
OK, this is good!

Please reply to all the forum threads.  People will be thrilled!

**Pull request merged**